File: easy/290. Word Pattern/290. Word Pattern.ts
function wordPattern(pattern: string, s: string): boolean {
    const sArr = s.split(" ")

    if (sArr.length !== pattern.length) {
        return false
    }

    const patternToWord = new Map()
    const wordToPattern = new Map()

    for (let i = 0; i < sArr.length; i++) {
        if ((patternToWord.has(pattern[i]) && patternToWord.get(pattern[i]) !== sArr[i]) || (wordToPattern.has(sArr[i]) && patternToWord.get(sArr[i]) !== pattern[i]))
            return false

        patternToWord.set(pattern[i], sArr[i])
        wordToPattern.set(sArr[i], pattern[i])
    }

    return true
};

File: easy/290. Word Pattern/290. Word Pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        s_arr = s.split()

        if len(pattern) != len(s_arr):
            return False

        pattern_to_word = {}
        word_to_pattern = {}

        for p, w in zip(pattern, s_arr):
            if pattern_to_word.get(p, w) != w or word_to_pattern.get(w, p) != p:
                return False
            pattern_to_word[p] = w
            word_to_pattern[w] = p

        return True
